src/httpserver/http_request.hpp

```diff
@@ -356,8 +359,23 @@ class http_request {
          }
      }
 
-     const std::string get_connection_value(const std::string& key, enum MHD_ValueKind kind) const;
-     const std::map<std::string, std::string, http::header_comparator> get_headerlike_values(enum MHD_ValueKind kind) const;
+     // Cache certain data items on demand so we can consistently return views
+     // over the data. Some things we transform before returning to the user for
+     // simplicity (e.g. query_str, requestor), others out of necessity (arg unescaping).
+     // Others (username, password, digested_user) MHD returns as char* that we need
+     // to make a copy of and free anyway.
+     struct data_cache {
+        std::string username;
+        std::string password;
+        std::string query_str;
+        std::string requestor_ip;
+        std::string digested_user;
+        http::arg_map unescaped_mhd_args;
```

Comment:
I'd just call it unescaped_args (mhd just happens to be our current underlying framework, but we aren't necessarily bound to it)